### PR TITLE
Initial rough support for deducing generic arguments in a call to a generic function.

### PR DESCRIPTION
**Pull request by zygoloid**

### CODE REVIEW

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp]: I think it would help to have a brief comment on this struct, in particular highlighting that this is basically just `deductions` plus some convenience methods for adding items to it. If it's semantically important that `deductions` be processed in stack order, that would be worth mentioning too.

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp:47]: It surprises me that this isn't an error; it might be worth having a comment to explain it.

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp]: Optional: I'd find `AddAll` a slightly more self-explanatory name, because it conveys that we're not somehow adding the block as a unit, we're adding each element individually.

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp]: Optional: would `[[maybe_unused]]` work instead?

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp:97]: If it's not valid, does that mean there was an earlier error and we're proceeding on a best-effort basis, or does it just mean there was no enclosing specific? If the latter, should `enclosing_specific_id` be a `std::optional`? I haven't gotten a handle on when we wrap ID types in `std::optional` vs when we rely on the invalid ID.

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp]: The name `results` is vague, and makes it a little harder for me to follow the logic. Would something like `generic_arg_ids` work instead?

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp]:
```suggestion
  llvm::SmallVector<SemIR::InstId> results;
  llvm::SmallVector<Substitution> substitutions;

  // Copy any outer generic arguments from the specified instance and prepare to
  // substitute them into the function declaration.
  if (enclosing_specific_id.is_valid()) {
```

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp]:
```suggestion
  results.resize(context.inst_blocks()
                     .Get(context.generics().Get(generic_id).bindings_id)
                     .size(),
                 SemIR::InstId::Invalid);

  worklist.AddBlock(params_id, arg_ids, /*needs_substitution=*/true);

```

(This is very optional, I just find it slightly clearer to group together the code that's setting up `results` for the main loop)

File: toolchain/check/deduce.cpp
Comment (geoffromer) [inline toolchain/check/deduce.cpp:198]:
```suggestion
      CARBON_CHECK(i >= first_deduced_index);
      auto binding_id = context.inst_blocks().Get(
```

This diagnostic seems like it would be confusing if it referred to a parameter of the enclosing generic, and presumably that's not supposed to happen anyway.

File: toolchain/check/testdata/function/generic/redeclare.carbon
Comment (geoffromer) [inline toolchain/check/testdata/function/generic/redeclare.carbon:37]: This seems like a regression: the old errors make sense given how the compiler is resolving `F`, but the new error does not. Is this an artifact of the earlier redeclaration error, or would something like this have the same problem?

```
fn G(T:! type, U:! type) -> T*;
fn F(T:! type, U:! type) -> U* {
  return G(T);
}
```

File: toolchain/check/testdata/function/generic/deduce.carbon
Comment (geoffromer) [inline toolchain/check/testdata/function/generic/deduce.carbon]:
```suggestion
library "deduce_explicit";
```

File: toolchain/check/testdata/function/generic/deduce.carbon
Comment (geoffromer) [inline toolchain/check/testdata/function/generic/deduce.carbon:95]: This is probably future work, but it would really be nice if we could diagnose that `U` isn't deducible for any possible usage, without having to see a usage.

**Review by geoffromer — COMMENTED** [review]
FYI I haven't reviewed the generated SemIR yet; I'm [looking for guidance](https://discord.com/channels/655572317891461132/655578254970716160/1269002893000704112) about how best to do that.

Also FYI, I'm on vacation next week, so if we don't finish this today you may need to find another reviewer.

File: toolchain/check/testdata/function/generic/deduce.carbon
Comment (geoffromer) [inline toolchain/check/testdata/function/generic/deduce.carbon:20]: Would it be worthwhile to have a test that covers passing a symbolic rather than template constant to a generic parameter?

File: toolchain/check/testdata/class/generic/method_deduce.carbon
Comment (geoffromer) [inline toolchain/check/testdata/class/generic/method_deduce.carbon:146]: Should this `%.4` be `constants.%.4`? If not, I can't find what it's referring to.

**Review by geoffromer — COMMENTED** [review]
(Finished initial review)

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp]: Done. Your comment caused me to realize that I do actually care about the processing order a little bit; changed to process in depth-first order and turned this into a class given that it's now maintaining an invariant.

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp:47]: Added comment.

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp]: Done.

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp]: Done.

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp:97]: All our `Id` types have an `::Invalid` value that represents "absent", as if there's an implicit `optional` wrapping them all. We use the invalid state to mean "none" pretty heavily.

(For what it's worth, I'm not sure "valid" / "invalid" is the best terminology here, because it has nothing to do with program validity, and we have a separate `Error` value for some of our `Id` types as a placeholder for "couldn't form this value due to an error". Might be worth changing this to `::Absent` and `is_present()` or similar.)

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp]: Renamed to `result_arg_ids`. I find it useful for this to capture that it's the output of the algorithm, given that we may also have input generic args from the `enclosing_specific_id`. Could go with `result_generic_arg_ids` if that works better for you, though it's a bit long.

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp]: Done. For me this communicates the intent less well (the initial value of `results` and `substitutions` are being computed by the process described by the comment, as if they were return values of that process), but I can see that being subjective. Happy to defer to reviewer's preference :)

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp]: Thanks, I agree that's clearer. Done.

File: toolchain/check/deduce.cpp
Comment (zygoloid) [reply toolchain/check/deduce.cpp:198]: Yeah, good point. Changed to only check the deduced arguments.

File: toolchain/check/testdata/function/generic/redeclare.carbon
Comment (zygoloid) [reply toolchain/check/testdata/function/generic/redeclare.carbon:37]: This is a consequence of our performing deduction before we try converting the call arguments, and deduction just immediately failing if the arity doesn't match. I'm planning on revisiting this to improve the diagnostic; I've added a TODO to this test for now.

File: toolchain/check/testdata/function/generic/deduce.carbon
Comment (zygoloid) [reply toolchain/check/testdata/function/generic/deduce.carbon:95]: Added a TODO.

File: toolchain/check/testdata/function/generic/deduce.carbon
Comment (zygoloid) [reply toolchain/check/testdata/function/generic/deduce.carbon:20]: Sure, added.

File: toolchain/check/testdata/class/generic/method_deduce.carbon
Comment (zygoloid) [reply toolchain/check/testdata/class/generic/method_deduce.carbon:146]: Names in SemIR formatted types are by default in the `constants` scope, because type constants are almost always in that scope. (This can be confusing, but it does substantially reduce the verbosity. Maybe we should reconsider, perhaps using a shorter name for `constants`?)

Comment (geoffromer) [reply toolchain/check/testdata/class/generic/method_deduce.carbon:146]: I don't have a strong opinion about how to spell names that refer to constants, but what really threw me off here was the fact that the previous line uses the longer form `constants.%.4`. Is there some significance to that? If not, it would be nice to fix the inconsistency one way or the other, but presumably that's out of scope for this PR.

**Review by geoffromer — APPROVED** [review]

**Pull request merged**
